### DISCUSSION TITLE
Generic Voting Subsystem

### DISCUSSION THREAD

**Discussion by NotAFile** [piqueserver/piqueserver#665] [Feature Requests]
Very simple. You create a `Vote` class/object, give it some text and a success callback (and maybe optionally, a % threshold and fail callback) and submit it. Players vote, callback gets called.

**Comment by godwhoa** [comment]
Would it really clean up or DRY up votekick or votemap?

**Comment by NotAFile** [comment]
Well, this issue was made with two things in mind:
 - adding packets that allowed clients to vote directly instead of via chat
 - people wanting to easily add their own vote types

**Comment by damienflament** [comment]
I think this voting system may be **designed as a more generic communication system between the client and the server**.

Recently, I started playing NMRiH, and it ships with a voting system to switch map, kick player etc. It's a graphical menu using the mouse.
But some server use another interaction menu triggered by a command send via the chat. Then, the player use the keypad keys (0 to 9) to navigate and select an action.

We should prevents mixing those two types of interaction by designing a generic menu:

## Behavior
The menu content is simply sent to the client (handling submenus might be a plus) and the selected entry is sent back to the server. Then, the server performs the action (a callback) associated with this entry.

The menu shown to the client can be customized by the server depending on multiple parameters:
 - Is the player logged in ?
 - the player team etc.

## UI
The UI currently shown by the clients (AoS, OS, etc.) for the team and weapon selection can be reused.

## Triggering
As the menu content is sent by the server, it can be triggered using different ways:
 - a protocol packet sent by the client,
 - a command sent using the chat,
 - a requirement met by the player or the game (beginning/end of the round/game, score of the player/team, elapsed time since the beginning of the round/game, etc.),
 - an action triggered by another player/client (think about the vote menu, triggered by one player but shown to the other ones).

Some common menus might be triggered using a protocol packet (main server menu and the voting menu), other ones are server specific and are triggered using commands.

**Comment by NotAFile** [comment]
I'm not so sure about using the same look for this as for choosing the team. Having the screen right in your face while playing for e.g. a votekick would be really annoying.

Stock menus would probably be very helpful for translations.
